### ISSUE TITLE
Question about EntityManagerExtensions - Instantiate with GO 

### ISSUE THREAD

**Issue by enzi** [Unity-Technologies/EntityComponentSystemSamples#62]
Hey all!

The code in question is:
```  
unsafe public static Entity Instantiate(this EntityManager entityManager, GameObject srcGameObject)
        {
            if (entityManager.m_CachedComponentList == null)
                entityManager.m_CachedComponentList = new List<ComponentDataWrapperBase>();

            var components = (List<ComponentDataWrapperBase>)entityManager.m_CachedComponentList;
            srcGameObject.GetComponents(components);
            var count = components.Count;
            ComponentType* componentTypes = stackalloc ComponentType[count];

            for (var t = 0; t != count; ++t)
                componentTypes[t] = components[t].GetComponentType();

            var srcEntity = entityManager.CreateEntity(entityManager.CreateArchetype(componentTypes, count));
            for (var t = 0; t != count; ++t)
                components[t].UpdateComponentData(entityManager, srcEntity);

            return srcEntity;
        }
```

What is the usage of m_CachedComponentList? I'm not able to see any working caching mechanisms. No early out.
Does srcGameObject.GetComponents has an internal logic for resolving the components?
If so, how does this cache work with multiple different gameobjects?

As the need still persists to instantiate multiple gameobjects in a single frame like UILabels or something I'm still falling back to pools, which Joachim mentioned should not be needed anymore.

I also scrubbed the documentation but found no answer or statement how fast this should be or the state of hybrid pooling.

**Comment by amechtley** [comment]
Hi! I think the name is maybe a little misleading (and there's some other stuff worth cleaning up in this excerpt), but there are two basic paradigms of GameObject.GetComponents and related methods:

```
void GetComponents(List<T> results)
T[] GetComponents<T>()
```

The former variant will resize the buffer passed in if needed, clear its contents, and populate it with the results of the call, but otherwise client code has full control over the buffer's lifetime.

The latter variant will return an array, allocating a heap object with every call.

Basically, the purpose of something like m_CachedComponentList is just to allocate a reusable buffer one time. Its contents are not needed again after it is used here (and really ought to be cleared before this method exits, which I will fix), though they will be cleared again with subsequent calls into this method.

As for your questions about hybrid pooling, I recommend raising it instead in the [forum](https://forum.unity.com/forums/entity-component-system-and-c-job-system.147/), which is probably a better place to get input on topics like that.
